_published/0014-test-results-handling.md

```diff
@@ -0,0 +1,220 @@
++  Total       uint      `gorm:"not null" json:"total"`
++  Failed      uint      `gorm:"not null" json:"failed"`
++  Passed      uint      `gorm:"not null" json:"passed"`
++  Skipped     uint      `gorm:"not null" json:"skipped"`
++}
+
++type TestResult struct {
++  ArtifactID  uint              `gorm:"not null;index:testresult_idx_artifact_id" json:"artifactId"`
++  Artifact    *Artifact         `gorm:"foreignKey:ArtifactID;constraint:OnUpdate:RESTRICT,OnDelete:RESTRICT" json:"-"`
++  Name        string	        `gorm:"not null" json:"name"`
++  Message     string            `gorm:"nullable" json:"message"`
++  StartedOn   *time.Time        `gorm:"nullable;default:NULL;" json:"startedOn" format:"date-time"`
++  CompletedOn *time.Time        `gorm:"nullable;default:NULL;" json:"finishedOn" format:"date-time"`
++  Status      TestResultStatus  `gorm:"not null" enums:"Failed,Passed,Skipped"`
++}
+```
+```go
+// artifact.go
+// new
+type File struct {
+  name string
+  fileName string
+  data []bytes
+}
+// new, /build/{buildid}/test-results
+func (m ArtifactModule) putTestResultsHandler(c *gin.Context) {
+  // parseMultipartFormData to get the files
+  // for each file, parse and store in db
+}
+// new, /build/{buildid}/artifact/{artifactid}/test-results
+func (m ArtifactModule) getBuildArtifactTestResultsHandler(c *gin.Context) {
+  // fetch test results for the specified test
+}
+// new
+func parseMultipartFormData(c *gin.Context) []*File {
+  // ...
+}
+// new
+func parseTRX(file *File) []TestResult, TestResultSummary {
+  // ...
+}
+```
+
+The new endpoint `GET /build/{buildid}/test-results-summary` fetches
+all test result summaries for a build and constructs a json response
+that contains them, and has a summary of them, like:
+
+<details><summary>GET /build/42/test-results-summary</summary>
+
+```json
+{
+  "failed": 4,
+  "passed": 5,
+  "skipped": 0,
+  "artifacts": [
+    {
+      "artifactId": 123,
+      "filename": "MyResults1.trx",
+      "failed": 1,
+      "passed": 3,
+      "skipped": 0
+    },
+    {
+      "artifactId": 124,
+      "filename": "MyResults2.trx",
+      "failed": 3,
+      "passed": 2,
+      "skipped": 0
+    }
+  ]
+}
+```
+</details>
+
+### wharf-web
+
+wharf-web changes to use the new endpoints
+- `GET /build/{buildid}/test-results-summary`\
+  To get summary of all test result summaries for build.
+  
+- `GET /build/{buildid}/artifact/{artifactid}/test-results`\
+  To get test result details for build.
+  
+- `POST /build/{buildid}/test-results`\
+  To upload test result files for build.
+
+Deprecated endpoint
+- `GET /build/{buildid}/tests-results`
+
+There would also be a way to view a build's test results' details. [#17](https://github.com/iver-wharf/wharf-api/issues/17)
+
+## Compatibility
+
+Nothing comes to mind.
+
+## Alternative solutions
+
+Nothing comes to mind.
+
+## Future possibilities
+
+Nothing comes to mind.
+
+## Unresolved questions
+
+Nothing comes to mind.
```

Comment:
(Moving @Pikabanga's discussion over here)

> I think we should try to model this to support other test result formats too. TRX is just one of many formats.
> 
> Not that I don't think this database model won't work for other formats too, but it could be worth looking into some other common test result formats to see if nothing obvious has been missed.

I personally find this current proposal quite extensible. We could consider attachments or other test fail states (such as the [MSTest's "inconclusive" state](https://docs.microsoft.com/en-us/previous-versions/visualstudio/visual-studio-2013/ms244275(v=vs.120)) in contrast to "skipped") later if we want to. I'm OK with leaving that for whenever we want to start supporting something like Java's JUnit, JS Jest, Go's built in test runner, et.al.

> One thing to consider is to add a mandatory parameter to the upload test results endpoint, to specify format. That way we don't need to do any file extension checking or content parsing to determine the format.
>
> /.../
>
> Edit: Just to be clear, I think it's perfectly fine to only support .trx in version 1, but I think the design should be made with multiple formats supported in mind.

I disagree with having a mandatory parameter, but an optional one would be benefitial.

Such as
- `POST /build/{buildid}/test-results`: Guess based on file extension, or error if unknown
- `POST /build/{buildid}/test-results?format=TRX`: Try parse as TRX, no matter what the file extension is.

Implementation wise it would just be a hardcoded map.

Reply:
> Implementation wise it would just be a hardcoded map.

Sure, I was more thinking of scenarios where there are multiple formats that store the results as .xml or .json. Then you would have to parse the content to determine the format. That was the complexity I wanted to avoid with the format parameter.

But sure, we can do a best effort to determine type (like TRX based on file extension), and throw an error if we are unable to.

Reply:
Oh yea, 100% agree. Attempting to guess based on content just stinks unexpected behaviour and overall bad UX